plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,305 @@
+   define generators that use them. This is done with functions like `bindTyName` and `bindTmName`:
+   ```
+   genLam ma mb = do
+      x <- genFreshName "x"
+      sizeSplit 1 7 (maybe (genType Star) return ma)
+                    --      v--- LOOK HERE!
+                    (\ a -> bindTmName x a . noEscape $ genTerm mb) $ \ a (b, body) ->
+                    --      ^--- LOOK HERE!
+                    TyFun () a b, LamAbs () x a body)
+   ```
+-}
+
+-- | Generate a fresh name. See Note [Warning about generating fresh names].
+genFreshName :: String -> GenTm Name
+genFreshName s = head <$> genFreshNames [s]
+
+-- | Generate one fresh name per string in the input list.
+-- names don't overlap. See Note [Warning about generating fresh names].
+genFreshNames :: [String] -> GenTm [Name]
+genFreshNames ss = do
+  used <- getUniques
+  let i = fromEnum $ Set.findMax $ Set.insert (Unique 0) used
+      js = [ j | j <- [1..i], not $ Unique j `Set.member` used ]
+      is = js ++ take (length ss + 10) [i+1..]
+  is' <- liftGen $ QC.shuffle is
+  return [Name (fromString $ s ++ show j) (toEnum j) | (s, j) <- zip ss is']
+
+-- | See `genFreshName`
+genFreshTyName :: String -> GenTm TyName
+genFreshTyName s = TyName <$> genFreshName s
+
+-- | See `genFreshNames`
+genFreshTyNames :: [String] -> GenTm [TyName]
+genFreshTyNames ss = map TyName <$> genFreshNames ss
+
+-- | Generate a name that overlaps with existing names on purpose. If there
+-- are no existing names, generate a fresh name.
+genNotFreshName :: String -> GenTm Name
+genNotFreshName s = do
+  used <- Set.toList <$> getUniques
+  case used of
+    [] -> genFreshName s
+    _  -> liftGen $ elements [ Name (fromString $ s ++ show (unUnique i)) i | i <- used ]
+
+-- | Generate a fresh name most (a bit more than 75%) of the time and otherwise
+-- generate an already bound name. When there are no bound names generate a fresh name.
+genMaybeFreshName :: String -> GenTm Name
+genMaybeFreshName s = frequency [(3, genFreshName s), (1, genNotFreshName s)]
+
+-- | See `genMaybeFreshName`
+genMaybeFreshTyName :: String -> GenTm TyName
+genMaybeFreshTyName s = TyName <$> genMaybeFreshName s
+
+-- | Bind a type name to a kind and avoid capturing free type variables.
+bindTyName :: TyName -> Kind () -> GenTm a -> GenTm a
+bindTyName x k = local $ \ e -> e
+    { geTypes = Map.insert x k (geTypes e)
+    , geTerms = Map.filter (\ty -> not $ x `Set.member` setOf ftvTy ty) (geTerms e)
+    , geDatas = Map.delete x (geDatas e)
+    }
+
+-- | Bind type names
+bindTyNames :: [(TyName, Kind ())] -> GenTm a -> GenTm a
+bindTyNames = flip $ foldr (uncurry bindTyName)
+
+-- | Remember that we have generated a type name locally but don't bind it.
+-- Useful for non-recursive definitions where we want to control name overlap.
+registerTyName :: TyName -> GenTm a -> GenTm a
+registerTyName n = local $ \ e -> e { geUnboundUsedTyNames = Set.insert n (geUnboundUsedTyNames e) }
```

Comment:
It's *very* easy to shoot yourself in the foot with the state technique. I strongly recommend avoding it specifically because of how easy it is to forget to apply `unbindTyName` in the right places.

The solution to forgetting is of course to not allow unmatched bind-unbind but then you don't gain anything from having state as you're now using state to emulate reader - so why not just use reader?

Now, the thing you may want to do is to use `QuoteT` for generating uniques and still use `reader` for binders etc. That would make guaranteeing global uniqueness *when you want it* easier. Which reminds me that there really ought to be a  flag that tells the generators to generate only globally unique names for when that's what you want and to generate potentially overlapping names when that's what you want.